/*! This file is part of davids91/Rafko.
 *
 *    Rafko is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    Rafko is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with Rafko.  If not, see <https://www.gnu.org/licenses/> or
 *    <https://github.com/davids91/rafko/blob/master/LICENSE>
 */

syntax = "proto3";
option cc_enable_arenas = true;
option java_package = "org.rafko.rafko_net";
option java_outer_classname = "RafkoRafkoNet";
package rafko_net;

import "common.proto";

/** @brief      Network basically describes the connections a recurrent network has
 *              with the past runs. This enumeration is for a utility which is for
 *              testing purposes only. Recurrences described here affect
 *              only the rpc @build_network in the service @RafkoDeepLearning
 */
enum Network_recurrence{
  network_recurrence_unknown = 0;
  network_recurrence_to_self = 1;
  network_recurrence_to_layer = 2;
}

/** @brief      Takes input, assigns weights to them and processes
 *              the weighted sums with a bias, transfer function and a memory filter.
 */
message Neuron{
  /* index of the weight for the memory filter. 1.0: keep its previous value instead of the newly calculated one */
  TransferFunctions transfer_function_idx = 3; /* index of the transfer function this Neuron uses, empty is invalid */

  /**
   * Input weights shall only contain positive intervals,
   * each interval corresponding to an index in the @weight_table in the @RafkoNet
   * Every weight above inputs in @input_indices count as a bias, and shall behave
   * like it has an input of 1.
   */
  repeated IndexSynapseInterval input_weights = 10;

  /**
   * Input indices describes intervals denoting the neurons inputs.
   * Each positive interval corresponds to indices in the internal data of a @Neuron stored in the @SolutionSolver
   * Each negative interval corresponds to indices in the input data given to the @SolutionSolver
   */
  repeated InputSynapseInterval input_indices = 11;
}

/** @brief      A sparse net implementation containing the
 *              Neurons and optimization cache files
 */
message RafkoNet{
  uint32 input_data_size = 10; /* Number of inputs (floating point numbers) accepted by the Neural network */
  uint32 output_neuron_number = 11; /* Number of outputs the Neural network has */

  repeated Neuron neuron_array = 20; /* Array of Neurons the network has */
  repeated double weight_table = 21; /* Stores individual weights used by the Neurons */
}
